usecases/config/runtime/manager.go

```diff
@@ -0,0 +1,194 @@
+			Help: "Whether the last loading attempt of runtime config was success",
+		}),
+		configHash: promauto.With(r).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "weaviate_runtime_config_hash",
+			Help: "Hash value of the currently active runtime configuration",
+		}, []string{"sha256"}), // sha256 is type of checksum and hard-coded for now
+	}
+
+	// try to load it once to fail early if configs are invalid
+	if err := cm.loadConfig(); err != nil {
+		return nil, err
+	}
+
+	return cm, nil
+}
+
+// Run is a blocking call that starts the configmanager actor. Consumer probably want to
+// call it in different groutine. It also respects the passed in `ctx`.
+// Meaning, cancelling the passed `ctx` stops the actor.
+func (cm *ConfigManager[T]) Run(ctx context.Context) error {
+	return cm.loop(ctx)
+	// cm.log.Info("starting runtime config manager")
+
+	// // Cancelling the passed in context is graceful shutdown
+	// if err := cm.loop(ctx); !errors.Is(err, context.Canceled) {
+	// 	cm.log.Error("stopping runtime config manager due to error", "msg", err.Error())
+	// }
+
+	// cm.log.Info("stopping runtime config manager")
+	// return nil
+}
+
+// Config returns the current valid config if available. Once the config manager
+// is started without any error, consumer should be able to get **valid** config
+// via this api.
+func (cm *ConfigManager[T]) Config() (*T, error) {
+	cm.mu.RLock()
+	defer cm.mu.RUnlock()
+
+	if cm.currentConfig == nil {
+		return nil, ErrEmptyConfig
+	}
+
+	return cm.currentConfig, nil
+}
+
+// loadConfig reads and unmarshal the config from the file location.
+func (cm *ConfigManager[T]) loadConfig() error {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+
+	f, err := os.Open(cm.path)
+	if err != nil {
+		cm.lastLoadSuccess.Set(0)
+		return errors.Join(ErrFailedToOpenConfig, err)
+	}
+	defer f.Close()
+
+	b, err := io.ReadAll(f)
+	if err != nil {
+		cm.lastLoadSuccess.Set(0)
+		return errors.Join(ErrFailedToReadConfig, err)
+	}
+
+	hash := fmt.Sprintf("%x", sha256.Sum256(b))
+	if hash == cm.currentHash {
+		cm.lastLoadSuccess.Set(1)
+		return nil // same file. no change
+	}
+
+	cfg, err := cm.parse(b)
+	if err != nil {
+		cm.lastLoadSuccess.Set(0)
+		return errors.Join(ErrFailedToParseConfig, err)
+	}
+
+	cm.currentConfig = cfg
+	cm.currentHash = hash
+
+	cm.lastLoadSuccess.Set(1)
+	cm.configHash.Reset()
+	cm.configHash.WithLabelValues(hash).Set(1)
+
+	return nil
+}
+
+// loop is a actor loop that runs forever till config manager is stopped.
+// it orchestrates between "loading" configs and "stopping" the config manager
+func (cm *ConfigManager[T]) loop(ctx context.Context) error {
+	ticker := time.NewTicker(cm.interval)
+	defer ticker.Stop()
+
+	// SIGHUP handler to trigger reload
+	sighup := make(chan os.Signal, 1)
+	signal.Notify(sighup, syscall.SIGHUP)
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := cm.loadConfig(); err != nil {
+				cm.log.Error("loading runtime config every %s failed, using old config: %w", cm.interval, err)
```

Comment:
for this kind of background operation, it is usually a good idea to use a sampled logger as an error might introduce a lot of noise in the logs

Reply:
My expectation here is the loadConfig will be called every 2m or so. So it's not a super noisy even if it's failing for hours.

But true it can be noisy if say loading interval is in milliseconds. I will add a note and come back to it when doing more testing.